src/number.rs

```diff
@@ -1,47 +1,327 @@
-use num_bigint::{BigInt, Sign};
-use std::sync::Mutex;
+use std::{fmt, ops::AddAssign};
+
+use ark_ff::{
+    fields::{Field, Fp64, MontBackend, MontConfig},
+    BigInteger, PrimeField, Zero,
+};
+
+#[derive(MontConfig)]
+#[modulus = "18446744069414584321"]
+#[generator = "3"]
+pub struct GoldilocksBaseFieldConfig;
+pub type GoldilocksBaseField = Fp64<MontBackend<GoldilocksBaseFieldConfig, 1>>;
 
 /// The abstract type of numbers to be computed with.
-/// They have arbitrary precision, but need to be converted
-/// to a finite field element once we generate the column values.
-pub type AbstractNumberType = num_bigint::BigInt;
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Default, PartialOrd, Ord, Hash)]
+pub struct AbstractNumberType {
+    value: GoldilocksBaseField,
+}
 /// The type of polynomial degrees and indices into columns.
 pub type DegreeType = u64;
 
-pub fn abstract_to_degree(input: &AbstractNumberType) -> DegreeType {
-    match input.to_biguint().unwrap().to_u64_digits()[..] {
-        [] => 0,
-        [d] => d,
-        _ => panic!(),
+impl AbstractNumberType {
```

Comment:
I would use `AbstractNumberType` for integers and `FiniteFieldElement` or something for this instead.

Reply:
I changed to using three types:
`FieldElement` for field elements
`AbstractNumberType` for integers (used for masks etc), right now u128
`DegreeType` for row indices, still u64

It's not 100% clean because we do not have types in the expressions so we still implicitly convert from one to the other but it's a bit cleaner than before this change imo.